File: app/src/main/java/org/ole/planet/myplanet/utilities/NewsMapper.kt
package org.ole.planet.myplanet.utilities

import org.ole.planet.myplanet.model.NewsItem
import org.ole.planet.myplanet.model.RealmNews
import java.text.SimpleDateFormat
import java.util.*

object NewsMapper {
    private val dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())

    fun fromRealm(realmNews: RealmNews): NewsItem {
        return NewsItem(
            id = realmNews.id,
            _id = realmNews._id,
            _rev = realmNews._rev,
            userId = realmNews.userId,
            user = realmNews.user,
            message = realmNews.message,
            docType = realmNews.docType,
            viewableBy = realmNews.viewableBy,
            viewableId = realmNews.viewableId,
            avatar = realmNews.avatar,
            replyTo = realmNews.replyTo,
            userName = realmNews.userName,
            messagePlanetCode = realmNews.messagePlanetCode,
            messageType = realmNews.messageType,
            updatedDate = realmNews.updatedDate,
            time = realmNews.time,
            createdOn = realmNews.createdOn,
            parentCode = realmNews.parentCode,
            imageUrls = realmNews.imageUrls?.toList(),
            images = realmNews.images,
            labels = realmNews.labels?.toList(),
            viewIn = realmNews.viewIn,
            newsId = realmNews.newsId,
            newsRev = realmNews.newsRev,
            newsUser = realmNews.newsUser,
            aiProvider = realmNews.aiProvider,
            newsTitle = realmNews.newsTitle,
            conversations = realmNews.conversations,
            newsCreatedDate = realmNews.newsCreatedDate,
            newsUpdatedDate = realmNews.newsUpdatedDate,
            chat = realmNews.chat,
            isEdited = realmNews.isEdited,
            editedTime = realmNews.editedTime,
            sharedBy = realmNews.sharedBy,
            formattedUpdatedDate = formatDate(realmNews.updatedDate),
            formattedTime = formatDate(realmNews.time),
            formattedNewsCreatedDate = formatDate(realmNews.newsCreatedDate),
            formattedNewsUpdatedDate = formatDate(realmNews.newsUpdatedDate),
            isCommunityNews = realmNews.isCommunityNews
        )
    }

    private fun formatDate(dateInMillis: Long): String {
        return if (dateInMillis > 0) {
            dateFormat.format(Date(dateInMillis))
        } else {
            ""
        }
    }
}

File: flutter/lib/data/local/news_mapper.dart
import 'dart:convert';

import 'package:drift/drift.dart';

import '../../core/utils/json_utils.dart';
import 'app_database.dart';

/// Port of `VoicesRepositoryImpl.buildNewsFromJson`.
///
/// The local row id is the CouchDB `_id` for anything the server sent, so a
/// re-sync updates a post in place rather than duplicating it. Locally composed
/// posts get a generated id instead and adopt the server's on upload.
class NewsMapper {
  const NewsMapper._();

  static NewsEntriesCompanion? fromDoc(
    Map<String, dynamic> doc, {
    NewsRow? existing,
  }) {
    final docId = JsonUtils.getString('_id', doc);
    if (docId.isEmpty || docId.startsWith('_design/')) return null;

    final user = JsonUtils.getObject('user', doc);
    final nested = JsonUtils.getObject('news', doc);

    return NewsEntriesCompanion.insert(
      // `buildNewsFromJson` seeds `id = underscoreId` for a post it has not
      // seen before, and otherwise keeps the existing row's id — which for a
      // locally composed post is the generated one, so an upload followed by a
      // sync does not fork the row.
      id: existing?.id ?? docId,
      docId: Value(docId),
      rev: Value(JsonUtils.getStringOrNull('_rev', doc)),
      viewableBy: Value(JsonUtils.getStringOrNull('viewableBy', doc)),
      viewableId: Value(JsonUtils.getStringOrNull('viewableId', doc)),
      docType: Value(JsonUtils.getStringOrNull('docType', doc)),
      avatar: Value(JsonUtils.getStringOrNull('avatar', doc)),
      updatedDate: Value(JsonUtils.getLong('updatedDate', doc)),
      createdOn: Value(JsonUtils.getStringOrNull('createdOn', doc)),
      messageType: Value(JsonUtils.getStringOrNull('messageType', doc)),
      messagePlanetCode: Value(
        JsonUtils.getStringOrNull('messagePlanetCode', doc),
      ),
      replyTo: Value(JsonUtils.getStringOrNull('replyTo', doc)),
      parentCode: Value(JsonUtils.getStringOrNull('parentCode', doc)),
      user: Value(user == null ? null : jsonEncode(user)),
      userId: Value(
        user == null ? null : JsonUtils.getStringOrNull('_id', user),
      ),
      userName: Value(
        user == null ? null : JsonUtils.getStringOrNull('name', user),
      ),
      time: Value(JsonUtils.getLong('time', doc)),
      message: Value(JsonUtils.getStringOrNull('message', doc)),
      images: Value(jsonEncode(_listOf('images', doc))),
      labels: Value(
        _listOf('labels', doc).map((e) => e.toString()).toList(growable: false),
      ),
      viewIn: Value(jsonEncode(_listOf('viewIn', doc))),
      chat: Value(JsonUtils.getBool('chat', doc)),
      newsId: Value(
        nested == null ? null : JsonUtils.getStringOrNull('_id', nested),
      ),
      newsRev: Value(
        nested == null ? null : JsonUtils.getStringOrNull('_rev', nested),
      ),
      newsUser: Value(
        nested == null ? null : JsonUtils.getStringOrNull('user', nested),
      ),
      aiProvider: Value(
        nested == null ? null : JsonUtils.getStringOrNull('aiProvider', nested),
      ),
      newsTitle: Value(
        nested == null ? null : JsonUtils.getStringOrNull('title', nested),
      ),
      conversations: Value(
        jsonEncode(
          nested == null ? const [] : _listOf('conversations', nested),
        ),
      ),
      newsCreatedDate: Value(
        nested == null ? 0 : JsonUtils.getLong('createdDate', nested),
      ),
      newsUpdatedDate: Value(
        nested == null ? 0 : JsonUtils.getLong('updatedDate', nested),
      ),
      // Read from the *nested* object, matching `buildNewsFromJson`, even
      // though `News.createNews` writes it at the top level. The asymmetry is
      // upstream's; reproducing it keeps a synced post identical to Kotlin's.
      sharedBy: Value(
        nested == null ? null : JsonUtils.getStringOrNull('sharedBy', nested),
      ),
      // Read from the *nested* object, because that is where `serializeNews`
      // writes it. Reading the top level here — as this did — meant a reaction
      // uploaded from one device was never seen by another, and worse, the next
      // sync-in of that same document wrote `Value(null)` straight over the
      // local column, so a user's own reaction vanished on their next sync.
      reactions: Value(
        nested == null ? null : JsonUtils.getStringOrNull('reactions', nested),
      ),
      // Local-only fields: an in-flight attachment list and the edit marker
      // belong to this device and are never sent back by the server.
      imageUrls: Value(existing?.imageUrls ?? const []),
      isEdited: Value(existing?.isEdited ?? false),
      editedTime: Value(existing?.editedTime ?? 0),
    );
  }

  static List<dynamic> _listOf(String key, Map<String, dynamic> doc) {
    final value = doc[key];
    return value is List ? value : const [];
  }
}
